### DISCUSSION TITLE
01.02 Metrics. [MOVED BY Felipe]

### DISCUSSION THREAD

**Discussion by Felipe-RA** [rramosp/ai4eng.v1#217] [01 - INTRODUCCIÓN]
```

@soyudea

Tengo que hacer esto
LAB 01.02 - Metrics

def f_rmsle():
t3_actual = np.random.randint(80,size=15)+20
t3_predicted = np.random.randint(80,size=15)+20

print ("actual ", t3_actual)
print ("predicted", t3_predicted)
n=np.size(t3_predicted)

suma = 0
for a,b in zip(t3_predicted, t3_actual):
a = np.log(a+1) #calculo parcial logaritmo
b = np.log(b+1) #Calculo parcial logaritmo
suma = suma + (a-b)**2 #Actualizacion suma

rmsle = ((suma)/n)**0.5 #Calculo de la raiz cuadrada
return rmsle

rmsle =f_rmsle()

valorformated = "{:.3f}".format(rmsle)
print(valorformated)

Resultados
actual [78 83 33 86 25 74 62 55 42 57 20 23 49 36 95] #Valores actuales
predicted [27 92 81 86 79 67 78 59 33 42 37 61 81 72 31] #Valor predicho

0.658

ejecuto la linea de codigo siguiente

rmsle = 0.658

error

KeyError Traceback (most recent call last)
in
----> 1 student.submit_task(globals(), task_id="task_03");

Los calculos manuales coinciden con lo que calcula el algoritmo
```

**Comment by Felipe-RA** [comment]
Saludos soyudea

No me queda claro cuál es tu error, pues no me sueltas todo el traceback.

De la misma forma, tampoco me queda claro qué tienes en cada celda.

Por favor haz uso de los **bloques de código** o compartenos un enlace abierto para todos a una copia de tu notebook en colab.

¡Estamos al pendiente!

**Comment by soyudea** [comment]
enlace al googleColab

**https://colab.research.google.com/drive/1Fnr_LoM4S_UbG-DoatoKzsNN80hS4MOS?usp=sharing**

Task 3: Evaluation in New York City Taxi Trip Duration Kaggle Competition
Understand the data and the evaluation metric (Root Mean Squared Logarithmic Error, RMSLE) of the following Kaggle competition

https://www.kaggle.com/c/nyc-taxi-trip-duration/
Observe that this competition is a regression task as we are measuring the difference in prediction with respect to the actual.


For instance, the following model predictions and ground truth:

actual    [66 37 22]
predicted [79 51 67]
produce a RMSLE of 0.66 aprox.

Execute the following cell to generate the data from which you must compute the metric. You may compute the metric implementing python code, or manually, or copy/pasting the actual and predicted data in Excel, etc.

Challenge: For python use numpy function np.log or np.log1p

Observe that every time you execute the following cell, a different set of values is generated. You will have to compute the metric for the values that you see. If you run the cell again you will have to compute your metric value again.

Este es el codigo

Task 3: Evaluation in New York City Taxi Trip Duration Kaggle Competition
Understand the data and the evaluation metric (Root Mean Squared Logarithmic Error, RMSLE) of the following Kaggle competition

https://www.kaggle.com/c/nyc-taxi-trip-duration/
Observe that this competition is a regression task as we are measuring the difference in prediction with respect to the actual.

For instance, the following model predictions and ground truth:

actual    [66 37 22]
predicted [79 51 67]
produce a RMSLE of 0.66 aprox.


Codigo
def f_rmsle():
   t3_actual = np.random.randint(80,size=15)+20
   t3_predicted = np.random.randint(80,size=15)+20
   #t3_actual = [66, 37,22]
   #t3_predicted =[79, 51, 67]
   print ("actual   ", t3_actual)
   print ("predicted", t3_predicted)
   n=np.size(t3_predicted)
   print(n)
   suma = 0
   for a,b in zip(t3_predicted, t3_actual):
       a = np.log(a+1)
       b = np.log(b+1)
       suma = suma + (a-b)**2
    
   rmsle = ((suma)/n)**0.5
   return rmsle
rmsle =f_rmsle()

valorformated = "{:.3f}".format(rmsle)   
print(valorformated)


actual    [78 83 33 86 25 74 62 55 42 57 20 23 49 36 95]
predicted [27 92 81 86 79 67 78 59 33 42 37 61 81 72 31]
15
resultado = 0.658

se supone que debe resolver :

[![image](https://user-images.githubusercontent.com/122616111/218223480-2b520e68-7cf5-41b0-80a7-0bfe28accbb9.png)
](https://www.kaggle.com/competitions/nyc-taxi-trip-duration/overview/evaluation)

Felipe
No me gusta pedir ayuda, pero esto está empezando
Si quieres tienes la opción de intentar ayudar o no
Me da la impresión de que la genialidad no te permite entender a los terrenales

**Reply by Felipe-RA** [reply]
Saludos @soyudea 

Gracias por compartirnos tu notebook. 

El motivo de tu error lo podemos evidenciar en el siguiente **mensaje de error**: 

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-128-e6667d9c547b> in <module>
----> 1 student.submit_task(globals(), task_id="task_03");

2 frames
/content/local/lib/rlxmoocapi/session.py in submit_task(self, namespace, course_id, session_id, lab_id, task_id, user_id, display_html)

/content/local/lib/rlxmoocapi/session.py in submit_task(self, namespace, course_id, session_id, lab_id, task_id, user_id, display_html)

/content/local/lib/rlxmoocapi/session.py in <dictcomp>(.0)

KeyError: 't3_actual'
```

Fijate que al final del **traceback** (rastreo de la cadena de errores) tenemos el siguiente error:

`KeyError: 't3_actual'`

Los errores de tipo `KeyError` ocurren cuando no se encuentra una `llave` en una estructura de tipo `dict`, pero se intenta acceder a esta.  (Quien intenta acceder a esta es el calificador automático, pero no la encuentra por lo siguiente)

En tu caso este error ocurre porque al meter `t3_actual` y `t3_predicted` dentro del alcance (en inglés `scope`) de la función `f_rmsle` que creaste, es como estuvieses "escondiendo" estas variables, por lo que el calificador automático no las ve.

Para solucionarlo manteniendo tu estructura de función te recomiendo incluir las dos variables `t3_actual` y `t3_predicted` como parámetros de la función, como te muestro abajo:


## Como lo tienes
```
def f_rmsle(t3_actual,t3_predicted):
   #t3_actual = [66, 37,22]
   #t3_predicted =[79, 51, 67]
   print ("actual   ", t3_actual)
   print ("predicted", t3_predicted)
   n=np.size(t3_predicted)
   print(n)
   suma = 0
   for a,b in zip(t3_predicted, t3_actual):
      a = np.log(a+1)
      b = np.log(b+1)
      suma = suma + (a-b)**2
    
   rmsle = ((suma)/n)**0.5
   return rmsle
```

## Cambio sugerido

```
def f_rmsle(t3_actual,t3_predicted):
   n=np.size(t3_predicted)
   print(n)
   suma = 0
   for a,b in zip(t3_predicted, t3_actual):
      a = np.log(a+1)
      b = np.log(b+1)
      suma = suma + (a-b)**2
    
   rmsle = ((suma)/n)**0.5
   return rmsle


t3_actual    = np.random.randint(80,size=15)+20
t3_predicted = np.random.randint(80,size=15)+20
print ("actual   ", t3_actual)
print ("predicted", t3_predicted)


rmsle =f_rmsle(t3_actual,t3_predicted)
```

Fijate que gracias a que defines las variables t3 como  **parámetros de la función** ( los paréntesis después de tu `def f_rmsle` ) puedes hacer uso de ellas sin necesidad de crearlas dentro de la propia función.

¡Me comentas como te va, hasta pronto!

PD: Con respecto a tu post anterior, lo moví porque estaba en un lugar incorrecto. De la misma manera, te pedí explicaciones extra pues no seguiste los #145 [Lineamientos para formular preguntas eficientemente] y para un tercero no familiarizado con tu código es difícil comprenderlo si no sigue algún formato.

No lo malinterpretes como no tener disposición de echarte una mano, ni mucho menos soy un genio. 😄
